sourcegraph/cody
# vscode: add feature flag provider

Adds feature flag provider used to evaluate feature flags and execute code conditionally based on flags value.
Currently works only for VSCode extensions users connected to dotcom instance. More context on it can be found in [this doc](https://docs.google.com/document/d/16IbXGLhHwXi7xnsSeHu3CPksV6NO1bHTNJ6B000IuJ0/edit#bookmark=id.9m8ljlh2hqs9) ([thread](https://docs.google.com/document/d/16IbXGLhHwXi7xnsSeHu3CPksV6NO1bHTNJ6B000IuJ0/edit?disco=AAAA22gAaU8)).

## Test plan
As we restrict feature flags evaluation only to dotcom users, the actual events data should be checked on dotcom instance `event_logs` table ([docs](https://docs.sourcegraph.com/dev/how-to/use_feature_flags#measure-the-effect-of-a-feature-flag)). 
During the development I tested the functionality in the following way:
- remove (comment out) `isDotCom` condition in `FeatureFlagProvider` methods
- run local Sourcegraph instance (`sg start`)
- log in to the local instance as admin and add a feature flag (e.g. `cody-embeddings-context-enabled`)
- sign it to the local Sourcegraph instance in Cody VSCode extension
- launch the VSCode extension from the VSCode Run&Debug menu
- perform a Cody completion action and check the `event_logs` table to see whether CodyVSCodeExtension logs include the correct feature flag value in `feature_flags` column
- double-check feature flag value using the GraphQL API
```
query Name {
  evaluateFeatureFlag(flagName: "cody-embeddings-context-enabled")
} 
```
